UPDATE: Rich Context HIPE (#151)
hyperledger/indy-hipe: text/0138-rich-schema-context/README.md

            "did:sov:JjmmTqGfgvCBnnPJRas6f8xT",
            "did:sov:3FtTB4kzSyApkyJ6hEEtxNH4",
            {
                "dct": "http://purl.org/dc/terms/",
                "rdf": "http://www.w3.org/1999/02/22-rdf-syntax-ns#",
                "rdfs": "http://www.w3.org/2000/01/rdf-schema#",
                "Driver": "did:sov:2mCyzXhmGANoVg5TnsEyfV8",
                "DriverLicense": "did:sov:36PCT3Vj576gfSXmesDdAasc",
                "CategoryOfVehicles": "DriverLicense:CategoryOfVehicles"
            }
        ]
    }
}
```

### Indy and Aries
The complete architecture for context objects involves three separate
repositories:
- `indy-node`: The code run by a validator node participating in an
instance of the indy ledger, e.g., the validators node in the Sovrin
network run `indy-node`. Changes to this code will enable context objects
to be written to and retrieved from an instance of indy.
- `indy-data-manager`: code which a client may use to communicate with
validator nodes in an indy network. Changes to this code will enable
context transaction requests to be sent to validator nodes.
`indy-data-manager` complies with the interface described by the

> `aries-data-registry-interface` and is built to plug in to the aries

Author: ```suggestion
`aries-verifiable-data-registry-interface` and is built to plug in to the aries
```

> ecosystem.

Author: ```suggestion
ecosystem. This means that any Aries-compatible `vdri` could call and use `indy-data-manager`.
```

> - `aries-dri`: This is the location of the `aries-data-registy-interface`.

Author: ```suggestion
- `aries-vdri`: This is the location of the `aries-verifiable-data-registy-interface`.
```

Changes to this code will enable users of any data registry with an

> `aries-dri`-compatible data manager to handle contexts.

Author: ```suggestion
`aries-vdri`-compatible data manager to handle contexts.
```

establishment of a common vocabulary.

### Stored on ledger

> Contexts will be written to the ledger. The identifier for a context

Reviewer: It would be great to define how the Context object is stored in State on the ledger side (the state is a Patricia-trie where key-value can be stored for efficient querying). In particular, what will be a unique key of the object, and what will be the value.

Author: This is an aspect of the process that we could certainly use some guidance with.
The format of the JSON is currently based on that of a DID Document, but it isn't necessary to actually write the DID Document to the ledger, as long as that is the format of the data as it is returned when the DID is dereferenced. We could format the data however it works best to do so, and use whatever ledger key makes sense, as long as the ID provided to and used by the end user is a DID that doesn't reflect the internal complexity.

> Contexts will be written to the ledger. The identifier for a context

Author: Should the DID be a fully-qualified DID (for example did:sov:id-string)?
Any ID  pointing to some other object should be fully qualified, but I don't think it is necessary that the ID for this object be fully qualified.

Should we require a particular format and content of id-string part of the DID?
Yes. The id-string should be the base-58 encoding of the full 256-bit hash of the value of the data object. 

The proposed format relies on a draft scheme, which may potentially cause problems for end users.
I agree. This is a concern for me as well, but there are few options for JSON canonicalization and none of them are standards.

If we use a fully-qualified DID with a prefix, and have a unique name and version for every object (unique within a network defined by a DID prefix), then can we use a hash of a name and version as id-string? It would allow to easily query not by ID, but by name and version as well.
I don't like this. The id-string should be a hash of the stored content. This allows it to serve as an additional assurance that the content requested is correct, and allows it to be directly comparable to content stored anywhere else, even in a cache. Any other identical content that someone wants to write to the ledger would have the same id-string (and therefore trying to write it to the ledger would result in an error). Using the hash of the name and version wouldn't give us any of these assurances.

In my opinion, query by name and version is certainly desirable, but doing so should be part of a future solution, not incorporated into the ID.

> Contexts will be written to the ledger. The identifier for a context

Reviewer 2: I think that we should reference which version of the JSON canonicalization we are using to help users be aware of the "moving target" nature.

 is a DID. This allows contexts to be resolved and dereferenced. 
 Because the context is an immutable content object, i.e. one that cannot
 be modified by a controller, the id-string of its DID is the base58
 representation of the SHA2-256 hash of the canonical form of the value of
 the data object of the content property. The canonicalization scheme we
 recommend is the IETF draft 
 [JSON Canonicalization Scheme (JCS).](https://tools.ietf.org/id/draft-rundgren-json-canonicalization-scheme-16.html)
 
### Example Context Object
```
{

>     "@context": [

Author: Some recent changes to the DID spec should mean that we can get rid of the additional context.

        "https://www.w3.org/ns/did/v1", 
        "did:sov:yfXPxeoBtpQABpBoyMuYYGx"
    ],
    "id": "did:sov:BmfFKwjEEA9W5xmSqwToBkrpYa3rGowtg5C54hepEVdA",
    "content":{

>         "type": "ctx",

Author: future query capabilities.

>         "name":"DriverLicense",

Author: There are two types of objects stored on the ledger, content and control. Control objects usually have keys and are controlled by some entity (standard DID Documents, Cred Defs, Revocation Registries).
Content objects do not have a controller, but instead consist of some immutable data.

For these immutable objects (like a context), we removed the controller property from a DID Document and added a content property. The content object contains a data field for the immutable data, a name and version for future query capabilities, and a hash object which provides the hash function used and the result of hashing the immutable data.

>         "name":"DriverLicense",

Reviewer: Can you please point to some up-to-date w3c docs describing this (if any)?

>         "name":"DriverLicense",

Reviewer 2: Immutable content objects not a part of the w3c spec; they are extensions.

        "version":"1.0",

>         "hash":{

Author: Do we need the hash here to know how to calculate the ID's DID id-string?
Yes, and to be able to verify the provided content.

Do I understand correctly that this is the hash of the "content" field? If so, I think we need to move hash out of "content" to make calculation easier.
This is the hash of the value of the data property.

>         "hash":{

Reviewer: So, two objects with the same `data` but different metadata (name, version, etc.) will have the same id-string. Is it intended? Why don't we include metadata into the value calculate the hash against?

>         "hash":{

Author: This is intended. If two objects have identical data, I do not see the use case for more than one of them to be written to the ledger.

>         "hash":{

Reviewer 2: The hash should be of the data only. The data should be equivalent to the data returned by other systems for the same content. The metadata is not returned by other systems storing the same object.

            "type": "SHA2-256",
            "value": "a005abbfcfaf7b0d703a7fc9fb86c8b71a33a10ef24d292984fc863c225205b9"
        },
        "data":{

>             "@context": [

Author: It should be included, as the id-string of the DID is a hash of the value of the data property.

>             "@context": [

Author: I'm open to suggestions for a better way to do this. We've formatted the data section to be a DID Document (in consideration of the fact that it can be retrieved by a DID).

>                 "did:sov:UVj5w8DRzcmPVDpUMr4AZhJ",

Reviewer: Should we store the Context itself as a dict (part of the request JSON), or as a JSON serialized string?

                "ver":1,
                "data":{
                    "@context": [
                        "https://www.w3.org/ns/did/v1", 
                        "did:sov:yfXPxeoBtpQABpBoyMuYYGx"
                    ],
                    "id": "BmfFKwjEEA9W5xmSqwToBkrpYa3rGowtg5C54hepEVdA",
                    "content":{
                        "type": "ctx",
                        "name":"DriverLicense",
                        "version":"1.0",
                        "hash":{
                            "type": "SHA2-256",
                            "value": "a005abbfcfaf7b0d703a7fc9fb86c8b71a33a10ef24d292984fc863c225205b9"
                        },
                        "data":{

>                             "@context": [

Reviewer: Can we try to reduce the number of nested objects here? Currently the context payload is under data/data/content/data/@context

data: Context.
{
    id: identifier the context,
    context: proposed context's value as JSON,
    name: proposed context's name string
    version: proposed context's version string,

>     ver: the version of the generic ledger context object template

Reviewer: What is a difference between `ver` and `version` here?

Author: It may be unnecessary to include both. `version` is the version of the context object. `ver` is the version of the JSON object written to the ledger that contains the context object.
